src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php

```diff
@@ -7,57 +7,182 @@
+use SprykerEco\Zed\Unzer\Business\Exception\UnzerException;
+use SprykerEco\Zed\Unzer\UnzerConstants;
 
-class UnzerMarketplaceCreditCardChargeProcessor implements UnzerChargeProcessorInterface
+class UnzerMarketplaceCreditCardChargeProcessor extends UnzerCreditCardChargeProcessor
 {
     /**
-     * @var \SprykerEco\Zed\Unzer\Business\Reader\UnzerReaderInterface
+     * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
+     * @param array<int> $salesOrderItemIds
+     *
+     * @throws \SprykerEco\Zed\Unzer\Business\Exception\UnzerException
+     *
+     * @return void
      */
-    protected $unzerReader;
+    public function charge(OrderTransfer $orderTransfer, array $salesOrderItemIds): void
+    {
+        $paymentUnzerTransfer = $this->unzerRepository->findPaymentUnzerByOrderReference($orderTransfer->getOrderReference());
+        if ($paymentUnzerTransfer === null) {
+            throw new UnzerException(sprintf('Unzer Payment not found for order reference %s', $orderTransfer->getOrderReference()));
+        }
+
+        $unzerPaymentTransfer = $this->unzerPaymentMapper
+            ->mapPaymentUnzerTransferToUnzerPaymentTransfer($paymentUnzerTransfer, new UnzerPaymentTransfer());
+        $unzerPaymentTransfer->setUnzerKeypair($this->getUnzerKeyPair($unzerPaymentTransfer));
+
+        $paymentUnzerOrderItemCollectionTransfer = $this->unzerRepository
+            ->getPaymentUnzerOrderItemCollectionByOrderId($unzerPaymentTransfer->getOrderId());
+
+        $authIdsGroupedByParticipantId = $this->groupAuthorizeIdsIndexedByParticipantIds($paymentUnzerTransfer);
+        $orderItemsGroupedByParticipantId = $this->getOrderItemsGroupedByParticipantId($paymentUnzerOrderItemCollectionTransfer, $orderTransfer, $salesOrderItemIds);
+        foreach ($orderItemsGroupedByParticipantId as $participantId => $itemCollectionTransfer) {
+            $unzerChargeTransfer = $this->createUnzerCharge($unzerPaymentTransfer, $itemCollectionTransfer, $authIdsGroupedByParticipantId[$participantId]);
+            $unzerChargeTransfer = $this->addExpensesToMarketplaceUnzerChargeTransfer(
+                $unzerChargeTransfer,
+                $orderTransfer,
+                $paymentUnzerOrderItemCollectionTransfer,
+                $participantId,
+            );
+
+            $this->unzerChargeAdapter->chargePartialAuthorizablePayment($unzerPaymentTransfer, $unzerChargeTransfer);
+        }
+
+        $this->updatePaymentUnzerOrderItemEntities($paymentUnzerOrderItemCollectionTransfer, $salesOrderItemIds);
+    }
 
     /**
-     * @var \SprykerEco\Zed\Unzer\Business\Payment\Mapper\UnzerPaymentMapperInterface
+     * @param \Generated\Shared\Transfer\PaymentUnzerTransfer $paymentUnzerTransfer
+     *
+     * @return array<string, string>
      */
-    protected $unzerPaymentMapper;
+    protected function groupAuthorizeIdsIndexedByParticipantIds(PaymentUnzerTransfer $paymentUnzerTransfer): array
```

Comment:
```suggestion
    protected function getAuthorizeIdsIndexedByParticipantIds(PaymentUnzerTransfer $paymentUnzerTransfer): array
```

https://spryker.atlassian.net/wiki/spaces/RFC/pages/2099413078/Accepted+RFC+Naming+of+methods+convention+update.+Getting+indexed+grouped+collections.